[4.1.0][CC] Update API Authentication docs (#5656)
wso2/docs-apim: en/docs/deploy-and-publish/deploy-on-gateway/choreo-connect/security/api-authentication/oauth2-access-tokens.md

# OAuth 2.0 Authentication

> By default all APIs expect an access token during invocation, unless [security has been disabled]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/security/api-authentication/disabling-security/) for the API.

Reviewer: Shall we add something about OAuth in this introduction? This is too disconnected from the title for a reader.

Author: Yeah, it is.. I'll try to make it sound better.

> When API-M acts as the Control Plane for Choreo Connect, you can generate an access token and invoke the API by following the steps in [here]({{base_path}}/consume/invoke-apis/invoke-apis-using-tools/invoke-an-api-using-the-integrated-api-console/). This is an access token retrieved from the Resident Key Manager packaged in an API Manager pack (of profile `control-plane`). 

Reviewer: ```suggestion
When API-M acts as the Control Plane for Choreo Connect, you can generate an access token and invoke the API by following the steps [here]({{base_path}}/consume/invoke-apis/invoke-apis-using-tools/invoke-an-api-using-the-integrated-api-console/). This is an access token retrieved from the Resident Key Manager packaged in an API Manager pack (of profile `control-plane`). 
```

## Access Token (JWT) Validation

> Choreo Connect can accept JWTs issued by a **trusted** Key Manager as a valid access token to invoke the APIs. JWT tokens are self-validated by Choreo Connect without validating it against the Authorization Server (Key Manager) that issued the JWT.

Reviewer: ```suggestion
Choreo Connect can accept JWT tokens issued by a **trusted** Key Manager as a valid access token to invoke the APIs. JWT tokens are self-validated by Choreo Connect without validating them against the Authorization Server (Key Manager) that issued the JWT.
```

-   **Issuer(iss)** - The issuer claim is a mandatory claim. Choreo Connect validates the **iss** claim present in the JWT against the issuer provided by the control plane or configured in config.toml.
-   **Subject(sub)** - The subject claim is also a mandatory claim in a token. Choreo Connect uses the value of the **sub** claim as the user of the secured API.
-   **Expiry time(exp)** - "exp" claim is also a mandatory claim. Choreo Connect validates the validity period of the token using the **exp** claim

## JWT Validation Configuration

When [Control Plane is enabled]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/concepts/apim-as-control-plane/), configurations for all Key Managers are fetched from API-M. For [Standalone Mode]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/concepts/as-a-standalone-gateway/), trusted Key Managers can be configured in `config.toml`. Refer to [Configuring an External Key Manager]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/security/api-authentication/configuring-an-external-key-manager/) to learn more.

> Following is an example standalone mode configuration for a trusted Key Manager. When Choreo Connect has API-M as the Control Plane, these configurations will be overridden with the Key Manager configurations retrieved from API Manager if the issuers (Key Managers) are identical. Furthermore, if the corresponding Key Manager is removed from the API Manager Admin portal, the token service added from the configuration will be used.

Reviewer 2: shall we elaborate/explain this more `if the issuers (Key Managers) are identical`, if possible

Author: Can you give an example?

Reviewer 2: I mean reader may find it difficult to understand what you mean by this phrase `if the issuers (Key Managers) are identical`. Like, is it the name under [[enforcer.security.tokenService]]

Author: Got the idea. Will update

When API-M acts as the Control Plane for Choreo Connect, you can generate an access token and invoke the API by following the steps [here]({{base_path}}/consume/invoke-apis/invoke-apis-using-tools/invoke-an-api-using-the-integrated-api-console/). This is an access token retrieved from the Resident Key Manager packaged in an API Manager pack (of profile `control-plane`).

## JWT validation configuration

- When [Control Plane is enabled]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/concepts/apim-as-control-plane/), configurations for all Key Managers are fetched from API-M. 
- For [Standalone Mode]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/concepts/as-a-standalone-gateway/), trusted Key Managers can be configured in `config.toml`. 

Refer to [Configuring an External Key Manager]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/security/api-authentication/configuring-an-external-key-manager/) to learn more.

> Following is an example standalone mode configuration for a trusted Key Manager. When Choreo Connect has API-M as the Control Plane, these configurations will be overridden with the Key Manager configurations retrieved from API Manager if the issuers (Key Managers) are identical. Which means the value for `issuer` in both configurations are the same. Furthermore, if the corresponding Key Manager is removed from the API Manager Admin portal, the token service added from the configuration will be used.

Reviewer: ```suggestion
The following is an example standalone mode configuration for a trusted Key Manager. When Choreo Connect has API Manager as the Control Plane, these configurations will be overridden with the Key Manager configurations retrieved from API Manager if the issuers (Key Managers) are identical. This means the value for `issuer` in both configurations is the same. Furthermore, if the corresponding Key Manager is removed from the API Manager Admin Portal, the token service added from the configuration will be used.
```

### Configure Multiple JWT issuers

>  There can be use cases in certain organizations, where multiple JWT issuers or key managers are used. In that case, Choreo Connect can be configured to work with JWTs issued by all of them. In the case of multiple JWT issuers are provided, Choreo Connect will sequentially check a JWT with all the available issuers. Valid JWT tokens will be cached and then expiry time only will be validated in the subsequent calls with the same token.

Reviewer 2: I feel that we need to add more explanation on how jwt issuer is selected when there are multiple given in the config

Author: You mean that we maintain a map based on the issuer, and we retrieve by reading the iss claim present in the token?

Reviewer 2: yes. how cc picks config for a particular token out of multiple configs.
